### ISSUE TITLE
Seamless access to cloud storage providers from Jupyter Lab and Argo workflows

### ISSUE THREAD

**Issue by gpelouze** [NaaVRE/NaaVRE#38]
These user stories focus on accessing cloud storage from NaaVRE. They consider two kinds of storage:
- the internal NaaVRE cloud storage (MinIO, which allows users and virtual lab developers to store the input and output of workflows)
- external cloud storage providers (anything that users might want to retrieve data from or save data to)

### User stories

(a) As a virtual lab user, I want to seamlessly access the NaaVRE cloud storage from the Jupyter Lab interface, so that I can add input files and retrieve output files for workflows that I run.

(b) As a virtual lab developer, I want to seamlessly access the NaaVRE cloud storage programmatically, so that I my code can read input and write output files.

(c) As a virtual lab user, I want to seamlessly access external cloud storage from the Jupyter Lab interface, so that I can add input files and retrieve output files for workflows that I run.

(d) As a virtual lab developer, I want to seamlessly access external cloud storage programmatically, so that I my code can read input and write output files.

#### Current limitations

Story (a) has very partial support. User can browse MinIO buckets in Jupyter Lab, but it is not possible to open files in Jupyter Lab, or transfer them to the Jupyter Lab storage. Furthermore, users have to generate their own access key from the MinIO interface, which creates a lot of friction (i.e. not seamless at all).

Story (b) has very partial support. Paragrammatical access is possible through the S3 API (eg. with the [minio](https://pypi.org/project/minio/) library in Python or [aws.s3](https://github.com/cloudyr/aws.s3) in R), but:
- it requires developers to learn how to use these libraries, which is not always intuitive
- it requires developers to write code that is specific to the storage MinIO/S3 interface, which means that they cannot easily reuse their code in other contexts
- it requires developers to manage authentication in their code. This means that users have to provide their own access keys, which creates friction (same limitation as for story a)

Story (c) is not supported by NaaVRE, but developers could still use libraries for specific providers. This as the same limitations as for story (b). 

Story (d) is not possible, because storage other than S3 cannot be mounted in NaaVRE. 

#### Alternatives

S3 browser and storage-specific libraries (see above).

### Features

- Automatically mount buckets from the NaaVRE cloud storage onto Jupyter Lab
- Automatically mount buckets from the NaaVRE cloud storage onto cells running in Argo workflows.
- Allow users to configure and mount external cloud storage onto Jupyter Lab.
- Automatically mount the external cloud storage configured by the users onto cells running in Argo workflows.

#### Requirements

1. Mount NaaVRE cloud storage onto all Jupyter Lab instances:
   - public bucket (`arn:aws:s3:::naa-vre-public`) with read-only permission
   - prefix in the user data bucket (`arn:aws:s3:::naavre-user-data/{username}`) with read-write permission
   - virtual-lab-specific buckets (e.g. `naa-vre-waddenzee-shared)
   - ensure that users cannot access other user's data
2. Mount NaaVRE cloud storage onto all workflow cells running in Argo (same sub-requirements as for 1.)
3. Ensure that the mount paths are clear and consistent:
   - the same bucket/prefix should have the same mount point in Jupyter and Argo
   - mount points should not depend on the username or the context (so that VL developers don't have to handle this in their code)
   - mount points should be intuitive (e.g. `/data/naavre-minio/public`, `/data/naavre-minio/my-data`)
4. Allow users to configure external cloud storage from the Jupyter Lab interface
5. Mount the user-configured storage onto Jupyter Lab
   - allow opening files with Jupyter Lab and copying to the Jupyter Lab storage
7. Mount the user-configured storage onto Argo workflow cells

**Comment by gpelouze** [comment]
Related issues:
- https://github.com/QCDIS/projects_overview/issues/124
- https://github.com/QCDIS/infrastructure/issues/198
- https://github.com/QCDIS/NaaVRE-deployer/issues/17
- https://github.com/NaaVRE/NaaVRE-helm/issues/7

**Comment by gpelouze** [comment]
Features will / won't do:


- Automatically mount buckets from the NaaVRE cloud storage onto Jupyter
  → Done (we mount the buckets)

- Automatically mount buckets from the NaaVRE cloud storage onto cells running in Argo workflows.
  → Difficult because Argo doesn't isolate users. One way to achieve this could be to make the Argo UI read-only. By doing so, the only way to submit workflows would be to go through the NaaVRE-workflow-service. That would allow us to add code to the template that mounts the buckets for the user that submits the workflow.
  - https://github.com/QCDIS/projects_overview/issues/423
  - Modify the NaaVRE-workflow-service to mount bucket volumes

- Allow users to configure and mount external cloud storage onto Jupyter Lab.
  → JupyterLab DataMount (pending issue resolution on their side for configuration persistence)

- Automatically mount the external cloud storage configured by the users onto cells running in Argo workflows.
  → Won't do. (Would require to write an equivalent of JL DataMount for Argo / arbitrary pods.)

**Comment by skoulouzis** [comment]
We should be careful with the users' buckets:
If user A creates a workflow, the workflow service will create a file in that user's bucket 
If user A shares their workflow with user B, then user B will get access to user's A bucket 


We should add the user's name on runtime

**Comment by KoenGreuell** [comment]
- [x] Create a separate issue for mounting other bucket s for specific users: https://github.com/NaaVRE/NaaVRE/issues/119
